tests/unit_tests/utils/file.cpp

```diff
@@ -7,11 +7,92 @@
 
 using namespace polybar;
 
+namespace {
+
+  auto expand_with_shell(string const& path) -> string {
+    auto cmd = command_util::make_command("echo " + path);
+    cmd->exec();
+
+    string result;
+    cmd->tail([&result](string output) {
+      result = std::move(output);
+    });
+
+    return result;
+  }
+
+}
+
 TEST(File, expand) {
-  auto cmd = command_util::make_command("echo $HOME");
-  cmd->exec();
-  cmd->tail([](string home) {
-      EXPECT_EQ(home + "/test", file_util::expand("~/test"));
-      });
+  // '/' is inserted to relative paths
+  EXPECT_EQ(file_util::expand("a"), "/a");
+
+  // '.' and '..' are handled
+  EXPECT_EQ(file_util::expand("/a/./b"), "/a/b");
+  EXPECT_EQ(file_util::expand("/a/b/../c"), "/a/c");
+  EXPECT_EQ(file_util::expand("/a/b/./././c"), "/a/b/c");
+
+  // consecutive '/'s are stripped
+  EXPECT_EQ(file_util::expand("/a/b////c"), "/a/b/c");
+
+  // trailing '/' is preserved
+  EXPECT_EQ(file_util::expand("a/"), "/a/");
+  EXPECT_EQ(file_util::expand("/a/./b/"), "/a/b/");
+
+  EXPECT_EQ(file_util::expand(""), "/");
+  EXPECT_EQ(file_util::expand("/"), "/");
+  EXPECT_EQ(file_util::expand("//"), "/");
+  EXPECT_EQ(file_util::expand("/../"), "/");
+  EXPECT_EQ(file_util::expand("/../.."), "/");
+  EXPECT_EQ(file_util::expand("/..////.."), "/");
+  EXPECT_EQ(file_util::expand("/.././.././.."), "/");
+  EXPECT_EQ(file_util::expand("."), "/");
+
+  // expand home directories
+  EXPECT_EQ(file_util::expand("~/test"), expand_with_shell("~/test"));
+  EXPECT_EQ(file_util::expand("./$HOME/test"), expand_with_shell("$HOME/test"));
+
+  EXPECT_NE(file_util::expand("~root/test"), "~root/test");
+  EXPECT_EQ(file_util::expand("~root/test"), expand_with_shell("~root/test"));
+
+  auto user_home = "~" + expand_with_shell("$USER") + "/test";
+  EXPECT_NE(file_util::expand(user_home), user_home);
+  EXPECT_EQ(file_util::expand(user_home), expand_with_shell(user_home));
+
+  // only leading tilde is expanded
+  EXPECT_EQ(file_util::expand("/tmp/~"), "/tmp/~");
+  EXPECT_EQ(file_util::expand("/tmp/~root"), "/tmp/~root");
+  EXPECT_EQ(file_util::expand("~/~"), expand_with_shell("~/~"));
+  EXPECT_EQ(file_util::expand("~/~"), expand_with_shell("$HOME/~"));
+
+  // HOME is expanded anywhere
+  EXPECT_EQ(file_util::expand("/tmp/$HOME"), expand_with_shell("/tmp$HOME"));
+  EXPECT_EQ(file_util::expand("/tmp/$HOME"), "/tmp" + expand_with_shell("~"));
 }
 
+TEST(File, dirname) {
```

Comment:
I don't think there's a lot of repetition between test cases, since each test has input specific to that particular function. I did refactor it so that the expected input and output are all declared first, and then the test runs for each pair. I think that accomplishes what you're describing without needing to set up a fixture or use the rest of the "parameterized test case" machinery.